runtime/oti/FFIUpcallThunkGenHelpers.hpp

```diff
@@ -323,4 +362,93 @@ AFI(I_8 *instructionPtr, I_32 sourceRegister, I_32 immediateInteger)
 
 }
 
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is smaller than or equal to 4096 bytes.
+ */
+static I_32
+generateSmallStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea, I_32 sourceRegister, I_32 destinationRegister)
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numLargeCopies = structSize / MAX_MVC_COPY_LENGTH;
+	const I_32 residue = structSize % MAX_MVC_COPY_LENGTH;
+
+	// Generate upto 16 MVC instructions to copy up 4096 bytes. Each MVC instruction in this sequence will copy 256 bytes.
+	for(int i = 0; i < numLargeCopies; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, i * MAX_MVC_COPY_LENGTH, MAX_MVC_COPY_LENGTH - 1, destinationRegister, i * MAX_MVC_COPY_LENGTH, sourceRegister);
+	}
+	// Generate an instruction to copy the last set of bytes (less than 256).
+	if (residue != 0) {
+		// MVC D1(L,B1),D2(B2)
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, numLargeCopies * MAX_MVC_COPY_LENGTH, residue - 1, destinationRegister, numLargeCopies * MAX_MVC_COPY_LENGTH, sourceRegister);
+	}
+	return totalInstructionSize;
+}
+
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is larger than 4096 bytes.
+ */
+static I_32
+generateLargeStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea, I_32 sourceRegister, I_32 destinationRegister)
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numFullSlots = structSize / MAX_MVC_COPY_LENGTH;
+	const I_32 residue = structSize % MAX_MVC_COPY_LENGTH;
+	const I_32 maxLargeCopies = MAX_MVC_DISPLACEMENT / MAX_MVC_COPY_LENGTH;
```

Comment:
please name maxLargeCopies more to the point ... at least it confused me when i read it.  it actually meant how many MVC instructions in an iteration of the copy loop?